src/Server/TCPHandler.cpp

```diff
@@ -764,7 +764,8 @@ void TCPHandler::processTablesStatusRequest()
     TablesStatusResponse response;
     for (const QualifiedTableName & table_name: request.tables)
     {
-        auto resolved_id = context_to_resolve_table_names->tryResolveStorageID({table_name.database, table_name.table});
+        auto resolved_id = context_to_resolve_table_names->tryResolveStorageID(
+            {table_name.database.empty() ? default_database : table_name.database, table_name.table});
```

Comment:
AFAIK `tryResolveStorageID` substitutes current database name if `table_name.database` is empty, so this change looks redundant

Reply:
@tavplubix thanks for taking a look. Please try running the integration test from this PR against a recent LTS. In 22.3 I'm seeing an error  ```There is no table ``.`test_data` on server: node1:9000```

Reply:
I trust you that some issue exists :) I mean that it's probably not a proper way to fix it.
Please run the test against recent master (22.8). If it passes then we should find the change that fixed it maybe backport it. If it fails then we should find the root cause of the issue (maybe `current_database` is not set yet at this point? or maybe we should fix `RemoteQueryExecutor` instead? or `AddDefaultDatabaseVisitor`?)

Reply:
Sure, that makes sense. Just tried the latest master, same issue.
```
Received exception from server (version 22.8.1):
Code: 279. DB::Exception: Received from localhost:9000. DB::NetException. DB::NetException: All connection tries failed. Log:

There is no table ``.`test_data` on server: localhost:9000
```
I'm fairly confident the issue is not in the way the root node sends the query, but in the way the query is received by the leaf executor.
Reasons I'm saying this are:
1) In 21.8 the same query works fine;
2) In a mixed cluster of 21.8 and 22.x the query works fine if it's received by the 22.x node;
3) My fix (which is on the receiving side of 22.x) works for a mixed cluster of 21.8 and 22.x nodes.

I'll try to do more digging to find a better place for the fix. In the meanwhile do you have a hunch where would be a good place to look? Thanks!

Reply:
I would check why `tryResolveStorageID` does not set database name to `current_database` as it should. Is `Context::setCurrentDatabase` already called at this point? For which context is it called and which context is used to resolve table ID? Also it may be related to #26864 (was merged in 21.9)

Reply:
Indeed, @tavplubix is right.

Here the culprit is inter-server mode, in this mode, since #26864, queries are executed w/o session context, and so they don't have default database

Looks like it is fine to use `default_database` always.

@vitlibar maybe you have an idea on a better/more correct fix?

Reply:
@tavplubix thank you for the pointers. I've updated the code to use a context. I understand the failure much better now but I can't say I follow the original intention of the code.

The bug (failure to resolve the default database) appears only if a remote node/shard definition has a `<secret>` set in it.

The following config will cause a failure:
```
<test_shard_localhost_default_database>
   <secret>foo</secret>
   <node>
     <host>localhost</host>
     <port>9000</port>
     <default_database>r0</default_database>
   </node>
 </test_shard_localhost_default_database>
```
However, removing `<secret>` makes the problem go away.

The difference is in the way the interserver request (one initiated by the distributed table) is handled.
The way the session context is created is affected by that as well.

First of all, `is_interserver_mode` is being set in [receiveHello() function](https://github.com/ClickHouse/ClickHouse/blob/master/src/Server/TCPHandler.cpp#L1069) in the following way:
`is_interserver_mode = (user == USER_INTERSERVER_MARKER) && password.empty();`
The effect of this is that when a <secret> is present in the shard config, is_interserver_mode is true, otherwise it's false.

[On the next line](https://github.com/ClickHouse/ClickHouse/blob/master/src/Server/TCPHandler.cpp#L1070), the session is created only for the non-interserver request. The default database is set for the session.
For interserver requests, the session is created (and then removed) later, when reading the query in [receiveQuery()
](https://github.com/ClickHouse/ClickHouse/blob/master/src/Server/TCPHandler.cpp#L1270) function.

The exception is thrown from [processTablesStatusRequests()
](https://github.com/ClickHouse/ClickHouse/blob/master/src/Server/TCPHandler.cpp#L767).
The function is called after the handshake (receiveHello) but before the query is read (receiveQuery). Since in the interserver mode the session doesn't exist before the query is received, the default server context is used which doesn't have the default database yet.


```
          ┌──────────┐                 ┌───────────────┐                  ┌────────────────────┐
          │ receive  │                 │processTable   │                  │ receiveQuery       │
<secret>  │ Hello    ├────────────────►│StatusRequest  ├───EXCEPTION!────►│                    │
          └──────────┘                 └───────────────┘                  └────────────────────┘
   is_interserver_mode=true         is_interserver_mode=true             is_interserver_mode=true
        session=nullptr                  session=nullptr                    session=makeSession()
                                      context=server.context()


          ┌──────────┐                 ┌───────────────┐                  ┌────────────────────┐
 no       │ receive  │                 │processTable   │                  │ receiveQuery       │
<secret>  │ Hello    ├────────────────►│StatusRequest  ├────SUCCESS──────►│                    │
          └──────────┘                 └───────────────┘                  └────────────────────┘
   is_interserver_mode=false        is_interserver_mode=false            is_interserver_mode=false
        session=makeSession()           session!=nullptr                     session!=nullptr
  sessionContext->setDefaultDB()    context=session->sessionContext()
```

Reply:
It would be helpful for me to understand the following to make a better change:
1. Why does the presense of <secret> affect how the interserver mode is handled.
2. In the interserver mode, why do we delay creation of the session context until a query is received?

Reply:
> Why does the presense of <secret> affect how the interserver mode is handled.

Because `<secret>` is the configuration parameter that enables this special interserver mode, see #13156. 

> In the interserver mode, why do we delay creation of the session context until a query is received?

I have no idea, let's ask @azat for more details.

Reply:
Thanks both for explaining, I missed @azat 's comment before writing my wall of text.

Reply:
> > In the interserver mode, why do we delay creation of the session context until a query is received?
>
>I have no idea, let's ask @azat for more details.

It is because in inter-server each query can have different credentials, it reads the client_info with the query and it has authentication details - https://github.com/ClickHouse/ClickHouse/blob/235649cb984371cb141f349f07959611062a5e12/src/Server/TCPHandler.cpp#L1305-L1352